### PR TITLE
Reliable aux msg

### PR DESCRIPTION
**Pull request by nvidianz**
Added support for reliable message to be used by XGBoost native implementation.

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

**Review by yanchengnv — APPROVED** [review]
LGTM

File: nvflare/apis/utils/reliable_message.py [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+
+# Topics for Reliable Message
+TOPIC_RELIABLE_REQUEST = "RM.RELIABLE_REQUEST"
+TOPIC_RELIABLE_REPLY = "RM.RELIABLE_REPLY"
+
+
+def _extract_result(reply: Shareable, target: str):
+    if not isinstance(reply, dict):
+        return None, None
+    result = reply.get(target)
```
Comment (YuanTingHsieh) [inline nvflare/apis/utils/reliable_message.py:50]: If we hinted that reply is a Shareable, should we use shareable getter methods?
Someday we might change it to not inherit from dict?

OR we should just type hint it to dict?

File: nvflare/apis/utils/reliable_message.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+def _status_reply(status: str):
+    return make_reply(rc=ReturnCode.OK, headers={HEADER_STATUS: status})
+
+
+def _error_reply(rc: str, error: str):
+    return make_reply(rc, headers={ReservedHeaderKey.ERROR: error})
+
+
+class _RequestReceiver:
+    def __init__(self, topic, request_handler_f, executor):
```
Comment (YuanTingHsieh) [inline nvflare/apis/utils/reliable_message.py:65]: what is the type of executor?

what is the signature of request_handler_f?

File: nvflare/apis/utils/reliable_message.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+                return receiver.process(request, fl_ctx)
+        else:
+            return make_reply(rc=ReturnCode.BAD_REQUEST_DATA)
+
+    @classmethod
+    def _receive_reply(cls, topic: str, request: Shareable, fl_ctx: FLContext):
+        tx_id = request.get_header(HEADER_TX)
+        receiver = cls._reply_receivers.get(tx_id)
+        if not receiver:
+            return make_reply(ReturnCode.OK)
```
Comment (YuanTingHsieh) [inline nvflare/apis/utils/reliable_message.py:208]: Why if not reply receiver is OK?

File: nvflare/apis/utils/reliable_message.py [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+                    # status should never be STATUS_NOT_RECEIVED, unless there is a bug in the receiving logic
+                    # STATUS_NOT_RECEIVED is only possible during "query" phase.
+                    break
+
+            # we didn't get a positive ack - wait a short time and re-send the request.
+            num_tries += 1
+            if num_tries > cls._max_retries:
+                # enough tries
+                return _error_reply(ReturnCode.COMMUNICATION_ERROR, f"Max send retries ({cls._max_retries}) reached")
+            time.sleep(cls._query_interval)
```
Comment (YuanTingHsieh) [inline nvflare/apis/utils/reliable_message.py:322]: should this be "resend_interval" or a fast checking period, then we use other "if" statement to determine if we need to resend?

File: nvflare/apis/utils/reliable_message.py [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+                topic=TOPIC_RELIABLE_REQUEST,
+                request=query,
+                timeout=timeout,
+                fl_ctx=fl_ctx,
+            )
+            ack, rc = _extract_result(ack, target)
+            if ack and rc != ReturnCode.COMMUNICATION_ERROR:
+                op = ack.get_header(HEADER_OP)
+                if op == OP_REPLY:
+                    # the ack is result itself!
```
Comment (YuanTingHsieh) [inline nvflare/apis/utils/reliable_message.py:353]: Do you think it make sense to have separate methods 
- one call "_send_request" which contains the previous while loop
- and one method call "_query_result" which contains this second while loop

File: nvflare/apis/utils/reliable_message.py [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+    _executor = None
+    _query_interval = 1.0
+    _max_retries = 5
+    _max_tx_time = 300.0  # 5 minutes
+    _reply_receivers = {}  # tx id => receiver
+    _tx_lock = threading.Lock()
+    _shutdown_asked = False
+
+    @classmethod
+    def register_request_handler(cls, topic: str, handler_f):
```
Comment (YuanTingHsieh) [inline nvflare/apis/utils/reliable_message.py:172]: might be good to add docstring about what "handler_f" needs to be

**Review by YuanTingHsieh — COMMENTED** [review]
Some feedbacks

File: nvflare/apis/utils/reliable_message.py (hunk 1 above)
Comment (nvidianz) [reply nvflare/apis/utils/reliable_message.py:50]: The reply is a dict, not shareable. Changed the hint.

File: nvflare/apis/utils/reliable_message.py (hunk 2 above)
Comment (nvidianz) [reply nvflare/apis/utils/reliable_message.py:65]: Added DocString.

File: nvflare/apis/utils/reliable_message.py (hunk 6 above)
Comment (nvidianz) [reply nvflare/apis/utils/reliable_message.py:172]: Added DocString.

File: nvflare/apis/utils/reliable_message.py (hunk 3 above)
Comment (nvidianz) [reply nvflare/apis/utils/reliable_message.py:208]: If no receiver found, it means the request is already processed so it returns OK. This happens when the query is sent before the result is received and the 2nd reply is received after the request is already finished.

File: nvflare/apis/utils/reliable_message.py (hunk 5 above)
Comment (yanchengnv) [reply nvflare/apis/utils/reliable_message.py:353]: Not sure what you  mean, but this is to try to make it more efficient by reducing the number of messages. So the query reply could be the result, or could also be an ack.

Comment (YuanTingHsieh) [reply nvflare/apis/utils/reliable_message.py:353]: I am suggesting we break this method into 2 smaller methods, each has their own for loop.

But that is just personal preference for better readability, logic wise this part LGTM

File: nvflare/apis/utils/reliable_message.py (hunk 4 above)
Comment (YuanTingHsieh) [reply nvflare/apis/utils/reliable_message.py:322]: For example:

'''
While True:

    If current time - previous time > resend interval:
         Do resend 

    time.sleep(0.1)
'''

Comment (nvidianz) [reply nvflare/apis/utils/reliable_message.py:322]: Changed to a fast loop.

File: nvflare/apis/utils/reliable_message.py (hunk 5 above)
Comment (nvidianz) [reply nvflare/apis/utils/reliable_message.py:353]: Good idea. Changed.

**Review by YuanTingHsieh — APPROVED** [review]
Thanks LGTM

**Pull request merged**